tests/e2e/slashing.go

```diff
@@ -374,106 +375,396 @@ func (suite *CCVTestSuite) TestOnRecvSlashPacketErrors() {
+		suite.Require().True(ubd.Entries[0].Balance.Equal(bondAmt.Quo(sdk.NewInt(2))))
+
+		// commit state on provider
+		suite.coordinator.CommitBlock(suite.providerChain)
+
+		// get the power after undelegate
+		validator = suite.getVal(suite.providerCtx(), valAddr)
+		powerAfterUndelegate = validator.GetConsensusPower(sdk.DefaultPowerReduction)
+		expectedPowerDelegated = (bondAmt.Quo(sdk.DefaultPowerReduction).Quo(sdk.NewInt(2))).Int64()
+		suite.Require().Equal(
+			powerBeforeDelegate+expectedPowerDelegated,
+			powerAfterUndelegate,
+			"unexpected power after undelegation; test: "+tc.name,
+		)
 
-	for _, tc := range testCases {
-		slashPacket := ccv.NewSlashPacketData(
-			abci.Validator{Address: tmValidator.Address, Power: tmValidator.VotingPower},
-			tc.vscID,
-			stakingtypes.Downtime,
+		// relay VSCPacket w/ undelegation from provider to consumer
+		relayAllCommittedPackets(
+			suite,
+			suite.providerChain,
+			suite.path,
+			ccv.ProviderPortID,
+			suite.path.EndpointB.ChannelID,
+			1,
+			"test: "+tc.name,
 		)
 
-		// slash
-		providerKeeper.HandleSlashPacket(suite.providerChain.GetContext(), suite.consumerChain.ChainID, *slashPacket)
+		// get the height when the consumer "applied" the undelegation
+		maturityTimes = consumerKeeper.GetAllPacketMaturityTimes(suite.consumerCtx())
+		suite.Require().Len(maturityTimes, 2, "unexpected number of maturity times; test: "+tc.name)
+		vscID = maturityTimes[1].VscId
+		hToVSCids = consumerKeeper.GetAllHeightToValsetUpdateIDs(suite.consumerCtx())
+		found = false
+		for _, hToVSCid := range hToVSCids {
+			if hToVSCid.ValsetUpdateId == vscID {
+				undelegateConsumerHeight = hToVSCid.Height
+				found = true
+				break
+			}
+		}
+		suite.Require().True(found, "cannot find height mapped to vscID; test: "+tc.name)
+
+		// create validator signing info to test slashing
+		valConsAddr, err := validator.GetConsAddr()
+		suite.Require().NoError(err, "test: "+tc.name)
+		providerSlashingKeeper.SetValidatorSigningInfo(
+			suite.providerChain.GetContext(),
+			valConsAddr,
+			slashingtypes.ValidatorSigningInfo{Address: valConsAddr.String()},
+		)
 
-		ubd, found := providerStakingKeeper.GetUnbondingDelegation(suite.providerChain.GetContext(), delAddr, valAddr)
+		// slash validator on consumer chain
+		consumerKey, found := providerKeeper.GetValidatorConsumerPubKey(suite.providerCtx(), suite.consumerChain.ChainID, valConsAddr)
 		suite.Require().True(found)
+		consumerAddr := utils.TMCryptoPublicKeyToConsAddr(consumerKey)
+		tc.slash(consumerAddr)
+
+		// increment time so that the unbonding period ends on the provider
+		incrementTime(suite, providerUnbondingPeriod)
 
-		isUbdSlashed := ubdBalance.GT(ubd.Entries[0].Balance)
-		suite.Require().True(tc.expSlash == isUbdSlashed)
+		tc.checkDelegatorBalance(delAddr, initBalance)
 
-		// update balance
-		ubdBalance = ubd.Entries[0].Balance
+		if i+1 < len(testCases) {
+			// reset suite to reset provider client
+			suite.SetupTest()
```

Comment:
This can be moved to the very beginning of the `for` loop.

Reply:
Yes, but then it will be called also in the first iteration and that's not needed. I don't think it save too much code.